File: postly-backend/accounts/middleware.py
from django.conf import settings

# Read by the Next.js middleware. Deliberately not httpOnly and deliberately
# not a credential: it carries no identity and grants nothing.
AUTH_HINT_COOKIE = "postly_auth"


class AuthHintCookieMiddleware:
    """
    Mirrors "somebody is logged in" into a cookie the frontend can read.

    The frontend needs a server-visible answer to "should I render the
    dashboard, or bounce to /login?" before React starts. Neither existing
    cookie can give it one:

    * `sessionid` is httpOnly, and — the part that actually bites — Django
      issues it to anonymous visitors too. allauth writes to the session
      during signup, so someone who has merely signed up carries a session
      cookie while being entirely unauthenticated. Treating its presence as
      "logged in" bounces them from /login to /dashboard, where the first
      API call 401s and sends them back to /login: an infinite loop.
    * `csrftoken` is set for everyone, logged in or not.

    So this sets a third cookie that tracks request.user, and clears it as
    soon as a request arrives unauthenticated — which includes the 401 from
    the frontend's own "who am I" call, so an expired session cleans up
    after itself on the way past.

    It is a hint for routing and nothing else. The API never reads it, and
    forging it gets you a dashboard shell with no data in it, because every
    endpoint still requires a real session.
    """

    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        response = self.get_response(request)

        user = getattr(request, "user", None)
        authenticated = bool(user and user.is_authenticated)
        present = request.COOKIES.get(AUTH_HINT_COOKIE) == "1"

        if authenticated and not present:
            response.set_cookie(
                AUTH_HINT_COOKIE,
                "1",
                max_age=settings.SESSION_COOKIE_AGE,
                domain=settings.SESSION_COOKIE_DOMAIN,
                secure=settings.SESSION_COOKIE_SECURE,
                httponly=False,
                samesite=settings.SESSION_COOKIE_SAMESITE,
            )
        elif not authenticated and present:
            response.delete_cookie(
                AUTH_HINT_COOKIE,
                domain=settings.SESSION_COOKIE_DOMAIN,
                samesite=settings.SESSION_COOKIE_SAMESITE,
            )

        return response

File: src/middleware.ts
import { NextResponse, type NextRequest } from "next/server";

/**
 * Keeps signed-out visitors from watching the dashboard render before its
 * first API call comes back 401, and keeps signed-in ones off /login.
 *
 * This is a convenience, not a security boundary. It reads a cookie and
 * nothing more: anyone can set it in devtools and walk straight past. What
 * actually protects the data is the API — every endpoint is IsAuthenticated
 * by default and every queryset is filtered by the requesting user, so a
 * forged cookie gets a dashboard shell with nothing in it. Never move a
 * real check up here.
 *
 * It reads `postly_auth`, not `sessionid`. Django hands out a session
 * cookie to anonymous visitors as well — allauth creates one during signup
 * — so "has a session cookie" is not the same question as "is logged in",
 * and answering the first bounces a signed-out visitor between /login and
 * /dashboard forever. `postly_auth` tracks request.user and is set and
 * cleared by AuthHintCookieMiddleware on the backend.
 *
 * The cookie reaches this middleware because cookies ignore ports, so :8000
 * and :3000 share them in development. Across subdomains in production it
 * needs SESSION_COOKIE_DOMAIN set on the backend.
 */

const AUTH_HINT_COOKIE = "postly_auth";
const SIGNED_OUT_ONLY = ["/login", "/signup"];

export function middleware(request: NextRequest) {
  const { pathname, search } = request.nextUrl;
  const hasSession = request.cookies.get(AUTH_HINT_COOKIE)?.value === "1";

  if (pathname.startsWith("/dashboard") && !hasSession) {
    const url = request.nextUrl.clone();
    url.pathname = "/login";
    url.search = `?next=${encodeURIComponent(pathname + search)}`;
    return NextResponse.redirect(url);
  }

  if (hasSession && SIGNED_OUT_ONLY.includes(pathname)) {
    const url = request.nextUrl.clone();
    url.pathname = "/dashboard";
    url.search = "";
    return NextResponse.redirect(url);
  }

  return NextResponse.next();
}

export const config = {
  matcher: ["/dashboard/:path*", "/login", "/signup"],
};
